### PR TITLE
OIDC: Add support for arbitrary claims for authorization subjects

### PR DESCRIPTION
**Pull request by vhdirk**
See https://github.com/eclipse/ditto/issues/512 for what is being implemented here

### CODE REVIEW

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultOAuthConfig.java
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultOAuthConfig.java:144]:
```suggestion
                    DefaultSubjectIssuerConfig.of(ConfigFactory.empty().withFallback(entry.getValue())));
```
This could/should work :)

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultSubjectIssuerConfig.java
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultSubjectIssuerConfig.java:29]:
```suggestion
public final class DefaultSubjectIssuerConfig implements SubjectIssuerConfig {
```
Please add `final` and javadoc for that public class.

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultSubjectIssuerConfig.java
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultSubjectIssuerConfig.java:45]: Please add javadoc for public methods.

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultSubjectIssuerConfig.java
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultSubjectIssuerConfig.java:99]:
```suggestion
    @Override
    public String toString() {
        return getClass().getSimpleName() + " [" +
                "issuer=" + issuer +
                ", authSubjectTemplates=" + authSubjectTemplates +
                "]";
    }
```
That toString() does not make too much sense - concatenating a string with a List.
Suggesting our "default" toString() here..

File: services/gateway/util/src/test/resources/oauth-test.conf
Comment (thjaeckle) [inline services/gateway/util/src/test/resources/oauth-test.conf:17]:
```suggestion
      issuer = "https://additional.google.com"
```

**Review by thjaeckle — CHANGES_REQUESTED** [review]
Finishing first round of review.
Looks already really promising, very cool. Thanks for the contribution 👍 

The parsing for JsonArray values is IMO a strong "must" as JWT claims may also be a JsonArrays of strings, I commented accordingly inline to your "TODO".
When you add this, please also provide another unit test case testing with a JWT containing a JsonArray as claim.

**Comment by thjaeckle** [comment]
@vhdirk you have one commit which did not include the "sign-off-footer" - but maybe you squash the commits again at the end, so we can worry about that later :)

**Comment by thjaeckle** [comment]
@vhdirk the remaining test error is because of test file `devops-test.conf` still having the "old" format of the issuers.

File: model/policies/src/main/java/org/eclipse/ditto/model/policies/SubjectIssuerConfig.java
Comment (vhdirk) [reply model/policies/src/main/java/org/eclipse/ditto/model/policies/SubjectIssuerConfig.java]: done

File: model/policies/src/main/java/org/eclipse/ditto/model/policies/SubjectIssuerConfig.java
Comment (vhdirk) [reply model/policies/src/main/java/org/eclipse/ditto/model/policies/SubjectIssuerConfig.java]: done

**Comment by vhdirk** [comment]
@thjaeckle I systematically forget to sign off my commits, so we'll have to squash this for sure :)

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/SubjectIssuerConfig.java
Comment (thjaeckle) [inline services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/SubjectIssuerConfig.java:18]: The `@link SubjectIssuer` won't work now because we are not in the same package and do not have that imported.
Suggesting:
```suggestion
 * Represents configuration for a {@link org.eclipse.ditto.model.policies.SubjectIssuer}, containing
```

File: services/gateway/security/src/main/java/org/eclipse/ditto/services/gateway/security/authentication/jwt/DittoJwtAuthorizationSubjectsProvider.java
Comment (vhdirk) [reply services/gateway/security/src/main/java/org/eclipse/ditto/services/gateway/security/authentication/jwt/DittoJwtAuthorizationSubjectsProvider.java]: done

File: services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultSubjectIssuerConfig.java
Comment (vhdirk) [reply services/gateway/util/src/main/java/org/eclipse/ditto/services/gateway/util/config/security/DefaultSubjectIssuerConfig.java]: done

File: services/gateway/security/src/test/resources/oauth-test.conf
Comment (thjaeckle) [reply services/gateway/security/src/test/resources/oauth-test.conf:16]: @vhdirk this is still pending - another config as well used "issuer:"

**Comment by thjaeckle** [comment]
Ah, some documentation on that would be nice as well: https://www.eclipse.org/ditto/installation-operating.html#openid-connect
There we also need to adjust the documented format anyways.

We could however also do that as follow-up task in order to not block the 2.0.0-M1 release.

**Comment by thjaeckle** [comment]
@vhdirk I just noticed that the `gateway.conf` still uses the old pattern (for the only configured default "google":
https://github.com/eclipse/ditto/blob/master/services/gateway/config/src/main/resources/gateway.conf#L176

We have to adjust that, otherwise I fear the Ditto gateway won't start.

**Comment by thjaeckle** [comment]
I fixed that in `gateway.conf` and resolved the conflict which was caused by that by merged master into this branch.

**Comment by vhdirk** [comment]
Thanks!

**Comment by thjaeckle** [comment]
@vhdirk I did try out the new feature in my local setup.
It basically works 👍 however .. when a `{{jwt:foo}}` placeholder cannot be resolved, it is added as unresolved placeholder to the list of SubjectIds, e.g. I used `"{{ jwt:sub }}/{{ jwt:scope }}"` instead of `"{{ jwt:sub }}/{{ jwt:scp }}"` which would have been correct and got at the `/whoami` HTTP API:
```json
{
    "defaultSubject": "mocky:bumlux",
    "subjects": [
        "mocky:bumlux",
        "mocky:bumlux/jwt:scope "
    ]
}
```

I'll look into where we can prevent this before merging the PR.

**Comment by vhdirk** [comment]
Wow, that was quick. I was looking at it out of curiosity and was just going to ask how you wanted this handled :)

**Review by thjaeckle — APPROVED** [review]

**Comment by thjaeckle** [comment]
Thanks again for this awesome contribution 🎉

**Pull request merged**
